### PR TITLE
Implement ftl::vector

### PR DESCRIPTION
**Pull request by BIGbadEL**

### CODE REVIEW

File: include/ftl/array.hpp [hunk 1]
```diff
@@ -212,7 +212,7 @@ template<typename T, std::size_t N>
 struct ftl::into_iterator_trait<ftl::array<T, N>, typename ftl::array<T, N>::iterator> {
   using iterator = typename array<T, N>::iterator;
 
-  [[nodiscard]] constexpr static auto into_iter(array<T, N> &arr) -> iterator {
+  [[nodiscard]] constexpr static auto into_iter(array<T, N> &arr) -> iterator {//TODO: ftl::array?
```
Comment (Blinkuu) [inline include/ftl/array.hpp]: ?

File: include/ftl/array_iterator.hpp [hunk 2]
```diff
@@ -27,6 +27,8 @@ class array_iterator final : public iterator_interface<array_iterator<Item, N>,
     : position_{ position }, begin_{ begin }, end_{ end } {
   }
 
+  //constexpr auto operator=(const array_iterator &) -> array_iterator & = default; //TODO: maybe dangerous
```
Comment (Blinkuu) [inline include/ftl/array_iterator.hpp]: Why this is commented out?

File: include/ftl/array_iterator.hpp [hunk 3]
```diff
@@ -84,6 +86,10 @@ class array_iterator final : public iterator_interface<array_iterator<Item, N>,
     return *this;
   }
 
+  auto predecrement_impl() const -> void {
```
Comment (Blinkuu) [inline include/ftl/array_iterator.hpp]: This is what decrement operator declaration looks like inside a class, as defined by the standard:
T& T::operator--();

File: include/ftl/array_iterator.hpp [hunk 4]
```diff
@@ -93,7 +99,7 @@ class array_iterator final : public iterator_interface<array_iterator<Item, N>,
   }
 
   [[nodiscard]] friend constexpr auto operator+=(const array_iterator<Item, N> &lhs, size_type n)
-    -> array_iterator<Item, N> & {
+    -> const array_iterator<Item, N> & {
```
Comment (Blinkuu) [inline include/ftl/array_iterator.hpp]: array_iterator return type was correct

File: include/ftl/array_iterator.hpp [hunk 5]
```diff
@@ -149,8 +155,8 @@ class array_iterator final : public iterator_interface<array_iterator<Item, N>,
 
 private:
   mutable size_type position_;
-  pointer const begin_;
-  pointer const end_;
+  pointer /* const */ begin_;
```
Comment (Blinkuu) [inline include/ftl/array_iterator.hpp]: Why this is commented out?

File: include/ftl/array_iterator.hpp [hunk 6]
```diff
@@ -149,8 +155,8 @@ class array_iterator final : public iterator_interface<array_iterator<Item, N>,
...
+  pointer /* const */ end_;
```
Comment (Blinkuu) [inline include/ftl/array_iterator.hpp]: Why this is commented out?

File: include/ftl/from_iterator_trait.hpp [hunk 7]
```diff
@@ -12,7 +12,7 @@ namespace ftl {
 template<typename T, typename U>
 struct from_iterator_trait {
   [[nodiscard]] constexpr static auto from_iter(const T &iter) -> U {
-    return { iter.begin(), iter.end() };
+    return U { iter.begin(), iter.end() };
```
Comment (Blinkuu) [inline include/ftl/from_iterator_trait.hpp:15]: This U is redundant.

File: include/ftl/iterator_interface.hpp [hunk 8]
```diff
@@ -227,6 +227,13 @@ class iterator_interface {
     return static_cast<Derived &>(*this).preincrement_impl();
   }
 
+  /**
+   * Decreases the iterator.
+   */
+  auto operator--() const -> void {
+    static_cast<const Derived &>(*this).predecrement_impl();
```
Comment (Blinkuu) [inline include/ftl/iterator_interface.hpp:234]: This should return decltype(auto) -> as as operator++().

File: include/ftl/vector_const_iterator.hpp [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+  using value_type = typename std::remove_cv_t<Item>;
+  using pointer = value_type *;
+  using reference = value_type &;
+  using const_pointer = const value_type *;
+  using const_reference = const value_type &;
+  using iterator_category = std::random_access_iterator_tag;
+  using size_type = std::size_t;
+  using std_vector_iterator = typename std::vector<value_type>::const_iterator;
+
+  constexpr vector_const_iterator() = default;
```
Comment (Blinkuu) [inline include/ftl/vector_const_iterator.hpp:23]: Is there a use case for default constructed iterator? I do not think so.

File: include/ftl/vector_const_iterator.hpp [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+  [[nodiscard]] constexpr auto deref_impl() -> reference {
+    return *current_;
+  }
+
+  [[nodiscard]] constexpr auto const_deref_impl() const -> const_reference {
+    return *current_;
+  }
+
+  auto preincrement_impl() const -> void {
```
Comment (Blinkuu) [inline include/ftl/vector_const_iterator.hpp]: T& T::operator++();

File: include/ftl/vector_const_iterator.hpp [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    ++current_;
+  }
+
+  auto predecrement_impl() const -> void {
```
Comment (Blinkuu) [inline include/ftl/vector_const_iterator.hpp]: T& T::operator--();

File: include/ftl/vector_const_iterator.hpp [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    --current_;
+  }
+
+  [[nodiscard]] constexpr auto count_impl() const -> size_type {
+    return static_cast<size_type>(end_ - begin_);
```
Comment (Blinkuu) [inline include/ftl/vector_const_iterator.hpp:75]: Maybe we could use std::distance here

File: include/ftl/vector_iterator.hpp [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+  [[nodiscard]] constexpr auto const_deref_impl() const -> const_reference {
+    return *current_;
+  }
+
+  auto preincrement_impl() const -> void {
+    ++current_;
+  }
+
+  auto predecrement_impl() const -> void {
```
Comment (Blinkuu) [inline include/ftl/vector_iterator.hpp]: T& T::operator--();

File: include/ftl/vector_iterator.hpp [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+  [[nodiscard]] constexpr auto deref_impl() -> reference {
+    return *current_;
+  }
+
+  [[nodiscard]] constexpr auto const_deref_impl() const -> const_reference {
+    return *current_;
+  }
+
+  auto preincrement_impl() const -> void {
```
Comment (Blinkuu) [inline include/ftl/vector_iterator.hpp]: T& T::operator++();

File: tests/CMakeLists.txt [hunk 15]
```diff
@@ -11,10 +11,12 @@ target_link_libraries(catch_main PRIVATE project_options)
 
 add_executable(tests 
   test_array.cpp
+  test_vector.cpp
   test_array_iterator.cpp
   test_enumerate_iterator.cpp
   test_list.cpp
   test_list_iterator.cpp
+        test_vector_iterator.cpp
```
Comment (Blinkuu) [inline tests/CMakeLists.txt]: Formatting grrr

File: tests/test_vector.cpp [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+  const auto ftl_vec = ftl::to_vector(vec);
+
+  REQUIRE(ftl_vec[0] == 1);
+  REQUIRE(ftl_vec[1] == 2);
+  REQUIRE(ftl_vec[2] == 3);
+  REQUIRE(ftl_vec[3] == 4);
+  REQUIRE(ftl_vec[4] == 6);
+}
+
+TEST_CASE(TEST_TAG "operator[]", TEST_TAG) {
```
Comment (Blinkuu) [inline tests/test_vector.cpp:19]: Maybe test address (pointers) validity as well

File: tests/test_vector.cpp [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+  REQUIRE(vec1.back() == 20);
+}
+
+TEST_CASE(TEST_TAG "emplace_back trivial type", TEST_TAG) {
+  ftl::vector<int> vec1 = { 5, 5, 5, 5, 5 };
+  vec1.emplace_back(20);
+  REQUIRE(vec1.back() == 20);
+}
+
+//TODO: test with custom class with non-trivial constructor
```
Comment (Blinkuu) [inline tests/test_vector.cpp]: Add task, because we should do this with every container.

File: tests/test_vector.cpp [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+}
+
+TEST_CASE(TEST_TAG "at out of range", TEST_TAG) {
+  const ftl::vector<int> vec1 = { 1, 2, 3, 4, 5 };
+  REQUIRE_THROWS(vec1.at(120));
+  REQUIRE(vec1.at(3) == 4);
+}
+
+//
+// TEST_CASE(TEST_TAG "iter", TEST_TAG) {
```
Comment (Blinkuu) [inline tests/test_vector.cpp]: ???

**Review by Blinkuu — CHANGES_REQUESTED** [review]

File: include/ftl/array_iterator.hpp (hunk 2 above)
Comment (BIGbadEL) [reply include/ftl/array_iterator.hpp]: Fixed

File: include/ftl/array_iterator.hpp (hunk 3 above)
Comment (BIGbadEL) [reply include/ftl/array_iterator.hpp]: fixed

File: include/ftl/array_iterator.hpp (hunk 4 above)
Comment (BIGbadEL) [reply include/ftl/array_iterator.hpp]: fixed

File: include/ftl/array_iterator.hpp (hunk 5 above)
Comment (BIGbadEL) [reply include/ftl/array_iterator.hpp]: fixed

File: include/ftl/vector_iterator.hpp (hunk 14 above)
Comment (BIGbadEL) [reply include/ftl/vector_iterator.hpp]: fixed

File: tests/CMakeLists.txt (hunk 15 above)
Comment (BIGbadEL) [reply tests/CMakeLists.txt]: clang format brrrrrrrrrr

File: tests/test_vector.cpp (hunk 16 above)
Comment (BIGbadEL) [reply tests/test_vector.cpp:19]: Added

File: tests/test_vector.cpp (hunk 17 above)
Comment (BIGbadEL) [reply tests/test_vector.cpp]: Done

File: tests/test_vector.cpp (hunk 18 above)
Comment (BIGbadEL) [reply tests/test_vector.cpp]: Fixed

**Review by Blinkuu — APPROVED** [review]

**Pull request merged**